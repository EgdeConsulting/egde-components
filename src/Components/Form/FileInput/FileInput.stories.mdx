import { Meta, Story, Canvas } from '@storybook/addon-docs';
import { DocsPageHeader, DocsPageBody } from 'Docs';
import { LibPath } from 'Types';
import { FileInput } from 'Components';

export const Template = (args) => <FileInput {...args} />;

export const componentName = 'FileInput';

<Meta
    title={LibPath.Form + componentName}
    component={FileInput}
    args={{
        label: 'Title',
        labelTextMargin: '0px',
        invalidText: '',
        captionText: '',
        margin: '0px',
        padding: '0px',
        width: '100%',
        minWidth: '0px',
        size: 'md',
        variant: 'default',
        uploadLabel: 'Drag and drop or click to upload',
        dropLabel: 'Drop',
        maxFileSizeLabel: 'Max file size is....',
        maxUploadSizeLabel: 'Max upload size is...',
        maxFileSize: 1000,
        maxUploadSize: 1000,
        disabled: false,
        showIcon: true,
        captionTextMargin: '0px',
        maxWidth: '100%',
    }}
/>

<DocsPageHeader
    title={componentName}
    description="This component is used to create an input which accepts files."
    addImportantNote={false}
    variant="stable"
/>

<Canvas withToolbar>
    <Story name="Base">{Template.bind({})}</Story>
</Canvas>

<DocsPageBody addStories={true} />

### Invalid upload

<Canvas withToolbar>
    <Story
        name="Invalid upload"
        args={{
            invalidText: 'Invalid upload',
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Without Icon

<Canvas withToolbar>
    <Story
        name="Without Icon"
        args={{
            showIcon: false,
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Disabled FileUpload

<Canvas withToolbar>
    <Story
        name="Disabled FileUpload"
        args={{
            variant: 'disabled',
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Small Size FileUpload

<Canvas withToolbar>
    <Story
        name="Small Size FileUpload"
        args={{
            size: 'sm',
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Medium Size FileUpload

<Canvas withToolbar>
    <Story
        name="Medium Size FileUpload"
        args={{
            size: 'md',
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>
