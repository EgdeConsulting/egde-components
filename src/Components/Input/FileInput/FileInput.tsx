/* eslint-disable */
import React, { ReactElement, useState } from 'react';
import { BaseInput } from 'Components';
import { FileUploadProps } from 'Types';
import Dropzone from 'react-dropzone';
import {
    Box,
    Center,
    List,
    ListItem,
    useMultiStyleConfig,
    IconButton,
} from '@chakra-ui/react';

import {
    UploadIcon,
    UploadDeleteIcon,
    UploadIconDisabled,
    UploadIconHover,
} from 'Icons';

import { FileInputStyles } from 'Styles/Components/FileInputStyles';

export const FileInput = (props: FileUploadProps): ReactElement => {
    const {
        label,
        labelTextMargin,
        invalidText,
        captionText,
        margin,
        padding,
        width,
        minWidth,
        size,
        variant,
    } = props;

    const fileInputStyles = useMultiStyleConfig('FileInput', { size, variant });

    const [totalUploadSizeReached, setTotalUploadSizeReached] =
        useState<boolean>(false);
    const [fileUploadSizeReached, setFileUploadSizeReached] =
        useState<boolean>(false);
    const [filesFromUser, setFilesFromUser] = useState<File[]>();
    const [hoverUploadBox, setHoverUploadBox] = useState(Boolean);

    const onDropAccepted = (acceptedFiles: any[]) => {
        let originalFileState = filesFromUser || [];
        let newFileState = (filesFromUser || []).concat(acceptedFiles);
        setFilesFromUser(newFileState);
        setTotalUploadSizeReached(false);

        if (newFileState) {
            let totalUploadSize = 0;
            totalUploadSize = newFileState
                .map((file: File) => file.size)
                .reduce((a: any, b: any) => {
                    return a + b;
                }, 0);

            if (totalUploadSize > props.maxUploadSize) {
                acceptedFiles = [];
                setFilesFromUser(originalFileState);
                setTotalUploadSizeReached(true);
            } else {
                setFileUploadSizeReached(false);
            }
        }
    };

    const onDropRejected = (rejectedFiles: any[]) => {
        rejectedFiles.forEach((rejected: any, index: number) => {
            if (rejected.errors[0].code === 'file-too-large') {
                setFileUploadSizeReached(true);
            }
        });
    };

    const onDeleteFile = (index: number) => {
        if (filesFromUser) {
            let array = filesFromUser;
            array.splice(index, 1);
            setFilesFromUser([...array]);

            if (filesFromUser.length === 0) {
                setFileUploadSizeReached(false);
                setTotalUploadSizeReached(false);
            }
        }
    };

    const getValidationText = (isDragActive: boolean) => {
        return (
            <Box
                __css={fileInputStyles}
                onMouseEnter={() =>
                    variant !== 'disabled' ?? setHoverUploadBox(true)
                }
                onMouseLeave={() =>
                    variant !== 'disabled' ?? setHoverUploadBox(false)
                }
            >
                <Center>
                    {!isDragActive &&
                    !totalUploadSizeReached &&
                    !fileUploadSizeReached
                        ? props.uploadLabel ??
                          'Dra og slipp filer her, eller klikk for å søke etter filer.'
                        : isDragActive
                        ? props.dropLabel ?? 'Slipp filer'
                        : totalUploadSizeReached
                        ? props.maxUploadSizeLabel ??
                          'Maks opplasting er ' +
                              (
                                  props.maxUploadSize /
                                  1024 /
                                  1024
                              ).toLocaleString() +
                              ' Mb'
                        : props.maxFileSizeLabel ??
                          'Maksstørrelse pr fil er ' +
                              (
                                  props.maxFileSize /
                                  1024 /
                                  1024
                              ).toLocaleString() +
                              ' Mb'}
                </Center>
                {props.showIcon && (
                    <Center marginTop={'0.5rem'}>
                        {hoverUploadBox === true ? (
                            <UploadIconHover />
                        ) : props.variant !== 'disabled' ? (
                            <UploadIcon />
                        ) : (
                            <UploadIconDisabled />
                        )}
                    </Center>
                )}
            </Box>
        );
    };

    return (
        <BaseInput
            label={label}
            labelTextMargin={labelTextMargin}
            invalidText={invalidText}
            captionText={captionText}
            margin={margin}
            padding={padding}
            width={width}
            minWidth={minWidth}
        >
            <Center>
                <Box
                    width={
                        size === 'sm'
                            ? FileInputStyles.sizes.sm.width
                            : FileInputStyles.sizes.md.width
                    }
                >
                    <Dropzone
                        onDropAccepted={onDropAccepted}
                        onDropRejected={onDropRejected}
                        maxSize={props.maxFileSize}
                        disabled={variant === 'disabled' ? true : false}
                    >
                        {({ getRootProps, getInputProps, isDragActive }) => {
                            return (
                                <div {...getRootProps()}>
                                    <input {...getInputProps()} />
                                    <Center>
                                        {getValidationText(isDragActive)}
                                    </Center>
                                </div>
                            );
                        }}
                    </Dropzone>
                    {filesFromUser && (
                        <List paddingTop={'0.5rem'}>
                            {filesFromUser.map((file, i) => (
                                <ListItem key={i}>
                                    {file.name +
                                        ' (' +
                                        Math.round(file.size / 1024) +
                                        ' Kb)'}

                                    <IconButton
                                        marginLeft={'0.5rem'}
                                        aria-label="Fjern fil"
                                        title="Fjern fil"
                                        size="xs"
                                        onClick={() => onDeleteFile(i)}
                                        icon={<UploadDeleteIcon />}
                                    />
                                </ListItem>
                            ))}
                        </List>
                    )}
                </Box>
            </Center>
        </BaseInput>
    );
};
