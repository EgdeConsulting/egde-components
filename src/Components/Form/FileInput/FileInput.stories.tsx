/* import { ComponentStory, ComponentMeta } from '@storybook/react';
import React from 'react';
import { FileInput } from './FileInput';
import { LibPath } from 'Types';

export default {
  title: LibPath.Form + 'FileInput' || 'undefined',
  component: FileInput,

  argTypes: {},
} as ComponentMeta<typeof FileInput>;

const Template: ComponentStory<typeof FileInput> = (args) => <FileInput {...args} />;

export const DeleteModalTest = Template.bind({});
 */
//  eslint-disable-next-line storybook/default-exports
