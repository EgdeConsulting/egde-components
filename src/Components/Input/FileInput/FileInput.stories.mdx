import { Meta, Story, Props, Canvas, Preview } from '@storybook/addon-docs';
import { FileInput } from './FileInput';
import { LibPath } from 'Types';

export const Template = (args) => <FileInput {...args} />;

<Meta
  title={LibPath.Form + 'FileInput'}
  component={FileInput}
  parameters={{
    toolbar: {
      title: { hidden: true },
      // zoom: { hidden: true },
      // eject: { hidden: true },
      // copy: { hidden: true },
      // fullscreen: { hidden: true },
      // 'storybook/background': { hidden: true },
      // 'storybook/viewport': { hidden: true },
    },
  }}
  args={{
    label: 'FileInput',
    labelTextMargin: '0px',
    invalidText: '',
    captionText: '',
    margin: '0px',
    padding: '0px',
    width: '100%',
    minWidth: '0px',
    size: 'md',
    variant: 'default',
    uploadLabel: 'Drag and drop or click to upload',
    dropLabel: 'Drop',
    maxFileSizeLabel: 'Max file size is....',
    maxUploadSizeLabel: 'Max upload size is...',
    maxFileSize: 1000,
    maxUploadSize: 1000,
    disabled: false,
    showIcon: true,
    captionTextMargin: '0px',
    maxWidth: '100%',
  }}
/>

# FileInput

<div
  style={{
    borderRadius: '999999px',
    display: 'inline-block',
    padding: '2px 8px',
    fontSize: '14px',
    lineHeight: '20px',
    fontWeight: '700',
    textAlign: 'center',
    letterSpacing: '0.25px',
    backgroundColor: '#138849',
    color: '#ffffff',
    marginBottom: '16px',
    marginRight: '3px',
  }}
>
  Stable
</div>

<div
  style={{
    borderRadius: '999999px',
    display: 'inline-block',
    padding: '2px 8px',
    fontSize: '14px',
    lineHeight: '20px',
    fontWeight: '700',
    textAlign: 'center',
    letterSpacing: '0.25px',
    backgroundColor: '#D2CB3F',
    color: '#ffffff',
    marginBottom: '16px',
    marginRight: '3px',
  }}
>
  Unstable
</div>

<div
  style={{
    borderRadius: '999999px',
    display: 'inline-block',
    padding: '2px 8px',
    fontSize: '14px',
    lineHeight: '20px',
    fontWeight: '700',
    textAlign: 'center',
    letterSpacing: '0.25px',
    backgroundColor: '#D23F47',
    color: '#ffffff',
    marginBottom: '16px',
    marginRight: '3px',
  }}
>
  Deprecated
</div>

<hr />

<Preview withToolbar>
  <Story name="FileInput">{Template.bind({})}</Story>
</Preview>

<Props story="FileInput" />

# Stories

<Canvas>
    <Story name="Invalid upload"
    args={{
      invalidText: 'Invalid upload',
    }}
    >{Template.bind({})}</Story>
</Canvas>

### Upload without icon
<Canvas>
    <Story name="Caption"
    args={{
      showIcon: false,
    }}
    >{Template.bind({})}</Story>
</Canvas>

### Disabled file upload
<Canvas>
    <Story name="Disabled"
    args={{
      variant: 'disabled',
    }}
    >{Template.bind({})}</Story>
</Canvas>

## Size variations

### Small
<Canvas>
    <Story name="Small"
    args={{
      size: 'sm',
    }}
    >{Template.bind({})}</Story>
</Canvas>

### Medium
<Canvas>
    <Story name="medium"
    args={{
      size: 'md',
    }}
    >{Template.bind({})}</Story>
</Canvas>

