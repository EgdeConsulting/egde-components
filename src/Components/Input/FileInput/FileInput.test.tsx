import React from 'react';
import { render } from '@testing-library/react';
import { FileInput } from './FileInput';

// Snapshot Test
describe('Snapshot FileInput', () => {
  it('should render correctly', () => {
    const tree = render(<FileInput
      label= 'FileInput'
      labelTextMargin= '0px'
      invalidText= ''
      captionText= ''
      margin= '0px'
      padding= '0px'
      width= '100%'
      minWidth= '0px'
      size= 'md'
      variant= 'default'
      uploadLabel= 'Drag and drop or click to upload'
      dropLabel= 'Drop'
      maxFileSizeLabel= 'Max file size is....'
      maxUploadSizeLabel= 'Max upload size is...'
      maxFileSize= {1000}
      maxUploadSize= {1000}
      disabled= {false}
      showIcon= {true}
      captionTextMargin= '0px'
      maxWidth= '100%'
        />
        );
    expect(tree).toMatchSnapshot();
  });
});


