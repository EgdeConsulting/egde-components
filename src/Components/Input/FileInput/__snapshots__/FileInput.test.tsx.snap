// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot FileInput should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
    <div>
      <div
        class="css-k456he"
      >
        <p
          class="chakra-text css-154u26z"
        >
          FileInput
        </p>
        <div
          class="css-1o2srd0"
        >
          <div
            class="css-gmuwbf"
          >
            <div
              class="css-6c3hwg"
            >
              <div
                role="presentation"
                tabindex="0"
              >
                <input
                  multiple=""
                  style="display: none;"
                  tabindex="-1"
                  type="file"
                />
                <div
                  class="css-gmuwbf"
                >
                  <div
                    class="css-0"
                  >
                    <div
                      class="css-gmuwbf"
                    >
                      Drag and drop or click to upload
                    </div>
                    <div
                      class="css-1p4hczz"
                    >
                      <svg
                        fill="none"
                        height="23"
                        viewBox="0 0 32 23"
                        width="32"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M25.8 9.05165C24.8933 4.45165 20.8533 0.998322 16 0.998322C12.1467 0.998322 8.8 3.18499 7.13333 6.38499C3.12 6.81165 0 10.2117 0 14.3317C0 18.745 3.58667 22.3317 8 22.3317H25.3333C29.0133 22.3317 32 19.345 32 15.665C32 12.145 29.2667 9.29165 25.8 9.05165ZM18.6667 12.9983V18.3317H13.3333V12.9983H9.33333L16 6.33165L22.6667 12.9983H18.6667Z"
                          fill="#01459E"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-k456he"
    >
      <p
        class="chakra-text css-154u26z"
      >
        FileInput
      </p>
      <div
        class="css-1o2srd0"
      >
        <div
          class="css-gmuwbf"
        >
          <div
            class="css-6c3hwg"
          >
            <div
              role="presentation"
              tabindex="0"
            >
              <input
                multiple=""
                style="display: none;"
                tabindex="-1"
                type="file"
              />
              <div
                class="css-gmuwbf"
              >
                <div
                  class="css-0"
                >
                  <div
                    class="css-gmuwbf"
                  >
                    Drag and drop or click to upload
                  </div>
                  <div
                    class="css-1p4hczz"
                  >
                    <svg
                      fill="none"
                      height="23"
                      viewBox="0 0 32 23"
                      width="32"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M25.8 9.05165C24.8933 4.45165 20.8533 0.998322 16 0.998322C12.1467 0.998322 8.8 3.18499 7.13333 6.38499C3.12 6.81165 0 10.2117 0 14.3317C0 18.745 3.58667 22.3317 8 22.3317H25.3333C29.0133 22.3317 32 19.345 32 15.665C32 12.145 29.2667 9.29165 25.8 9.05165ZM18.6667 12.9983V18.3317H13.3333V12.9983H9.33333L16 6.33165L22.6667 12.9983H18.6667Z"
                        fill="#01459E"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
